pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,167 @@
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should has status synced", "name", nsWObjRef.Name)
+					theSingletonState.Set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object should not have status synced", "name", nsWObjRef.Name)
+					theSingletonState.Unset(wObjIdentifier)
+				}
+				// reconcile should be out of the for loop over all bindings
+				// because the desired state for a single workload object must be determined over ALL bindings
+				if reconcile {
+					c.reconcileSingletonByWObj(ctx, wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func metav1GVR2schemaGVR(in metav1.GroupVersionResource) (out schema.GroupVersionResource) {
+	out = schema.GroupVersionResource{
+		Group:    in.Group,
+		Version:  in.Version,
+		Resource: in.Resource,
+	}
+	return
+}
+
+func (c *Controller) reconcileSingletonByWObj(ctx context.Context, wObjID util.ObjectIdentifier) error {
+	logger := klog.FromContext(ctx)
+
+	if !theSingletonState.Get(wObjID) {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &wObjID, emptyStatus, c.listers, c.wdsDynClient)
+	}
+
+	wsNameSuffix := strings.Join([]string{
+		wObjID.ObjectName.Namespace,
+		wObjID.ObjectName.Name,
+	}, "-")
+	// A (probably critical) question: How do I know which WEC is selected?
+	list, _ := c.workStatusLister.ByNamespace("cluster1").List(labels.Everything())
+	for _, obj := range list {
+		wsName := obj.(metav1.Object).GetName()
+		if !strings.HasSuffix(wsName, wsNameSuffix) {
+			continue
+		}
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return err
+		}
+		if status == nil {
+			return nil
+		}
+		logger.V(2).Info("Got status for workstatus", "name", wsName)
+		wsRef, _ := runtimeObjectToWorkStatusRef(obj)
+		return updateObjectStatus(ctx, &wsRef.sourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+	}
+	return nil
+}
+
+// To be improved
+func (c *Controller) reconcileSingletonByBdg(ctx context.Context, ref bindingRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(2).Info("Reconciling singleton status by binding", "name", string(ref))
+	return c.buildSingletonStateAndOptionallyReconcile(ctx, true)
+}
+
+func (c *Controller) reconcileSingletonByWs(ctx context.Context, ref singletonWorkStatusRef) error {
+	logger := klog.FromContext(ctx)
+	obj, _ := c.workStatusLister.ByNamespace(ref.wecName).Get(ref.name)
+	status, _ := util.GetWorkStatusStatus(obj)
+	if sync, ok := theSingletonState.wObjSync[ref.sourceObjectIdentifier]; !ok {
+		logger.V(2).Info("Not a singleton workload object, aborting", "sync", sync)
```

Comment:
Someone reading the debug log is going to wonder _what_ is being aborted. I think this message would be clearer if the ", aborting" were removed.